sdk/servicebus/Azure.Messaging.ServiceBus/TROUBLESHOOTING.md

```diff
@@ -0,0 +1,305 @@
+
+### How to browse scheduled or deferred messages
+
+Scheduled and deferred messages are included when peeking messages. They can be identified by the [ServiceBusReceivedMessage.State][MessageState] property. Once you have the [SequenceNumber][SequenceNumber] of a deferred message, you can receive it with a lock via the [ReceiveDeferredMessagesAsync][ReceiveDeferredMessages] method.
+
+When working with topics, you cannot peek scheduled messages on the subscription, as the messages remain in the topic until the scheduled enqueue time. As a workaround, you can construct a [ServiceBusReceiver][ServiceBusReceiver] passing in the topic name in order to peek such messages. Note that no other operations with the receiver will work when using a topic name.
+
+### How to browse session messages across all sessions
+
+You can use a regular [ServiceBusReceiver][ServiceBusReceiver] to peek across all sessions. To peek for a specific session you can use the [ServiceBusSessionReceiver][ServiceBusSessionReceiver], but you would need to obtain a session lock.
+
+## Troubleshoot processor issues
+
+### Autolock renewal is not working
+
+Autolock renewal relies on the system time to determine when to renew a lock for a message or session. If your system time is not accurate, e.g. your clock is slow, then lock renewal may not happen before the lock is lost. Ensure that your system time is accurate if autolock renewal is not working.
+
+### Processor appears to hang or have latency issues when using high concurrency
+
+This is often caused by thread starvation, particularly when using the session processor and using a very high value for [MaxConcurrentSessions][MaxConcurrentSessions] (e.g. > 50). The first thing to check would be to make sure you are not doing sync-over-async in any of your event handlers. Sync-over-async is an easy way to cause deadlocks and thread starvation. Even if you are not doing sync over async, any pure sync code in your handlers could contribute to thread starvation. If you've determined that this is not the issue, e.g. because you have pure async code, you can try increasing your [TryTimeout][TryTimeout]. This will relieve pressure on the threadpool by reducing the number of context switches and timeouts that may occur when using the session processor in particular. The default value for [TryTimeout][TryTimeout] is 60 seconds, but it can be set all the way up to 1 hour. Try setting it to something like 5 minutes as a starting point, and iterate from there. If none of this works, you may simply need to scale out to multiple instances. Reduce the concurrency in your application, but run the application on multiple instances to achieve the desired overall concurrency.
```

Comment:
I'd also consider mentioning that even with pure asynchronous code, there's no fairness in the scheduling algorithm.  If there's a high number of continuations waiting, some work/sessions/entitites/whatever may appear to stall while others do not.

Reply:
> I wonder if we should consider representing the "high concurrency" scenario as a factor of the available cores on the host rather than an absolute number. For a 2-core machine, 50 is crazy. For a 32-core machine, 50 is quite reasonable.

That concurrency can work okay. We throttle when accepting sessions based on the number of cores already. I plan to revise this section based on the recent developments re: the service bug.